contracts/adapters/AutoRepay.sol

```diff
@@ -0,0 +1,304 @@
+    if (repayParams.useFlashloan) {
+      bytes memory params = abi.encode(repayParams, permitSignature, msg.sender);
+      address[] memory assets = new address[](1);
+      assets[0] = repayParams.debtAsset;
+      uint256[] memory amounts = new uint256[](1);
+      amounts[0] = repayParams.debtRepayAmount;
+      uint256[] memory modes = new uint256[](1);
+      modes[0] = 0;
+      LENDING_POOL.flashLoan(address(this), assets, amounts, modes, repayParams.user, params, 0);
+    } else {
+      DataTypes.ReserveData memory debtReserveData = _getReserveData(repayParams.debtAsset);
+      uint256 amountToRepay;
+      {
+        address debtToken = DataTypes.InterestRateMode(repayParams.rateMode) ==
+          DataTypes.InterestRateMode.STABLE
+          ? debtReserveData.stableDebtTokenAddress
+          : debtReserveData.variableDebtTokenAddress;
+        uint256 currentDebt = IERC20(debtToken).balanceOf(repayParams.user);
+        amountToRepay = repayParams.debtRepayAmount <= currentDebt
+          ? repayParams.debtRepayAmount
+          : currentDebt;
+      }
+      uint256 maxCollateralToSwap = repayParams.collateralAmount;
+      if (amountToRepay < repayParams.debtRepayAmount) {
+        maxCollateralToSwap = maxCollateralToSwap.mul(amountToRepay).div(
+          repayParams.debtRepayAmount
+        );
+      }
+      repayParams.collateralAmount = maxCollateralToSwap;
+      repayParams.debtRepayAmount = amountToRepay;
+      _doSwapAndPullWithFee(repayParams, permitSignature, msg.sender, 0);
+
+      // Repay debt. Approves 0 first to comply with tokens that implement the anti frontrunning approval fix
+      IERC20(repayParams.debtAsset).safeApprove(address(LENDING_POOL), 0);
+      IERC20(repayParams.debtAsset).safeApprove(address(LENDING_POOL), repayParams.debtRepayAmount);
+      LENDING_POOL.repay(
+        repayParams.debtAsset,
+        repayParams.debtRepayAmount,
+        repayParams.rateMode,
+        repayParams.user
+      );
+    }
+    _checkHealthFactorInRange(repayParams.user);
+  }
+
+  function _doSwapAndPullWithFee(
+    RepayParams memory repayParams,
+    PermitSignature memory permitSignature,
+    address caller,
+    uint256 premium
+  ) internal {
+    address collateralATokenAddress = _getReserveData(repayParams.collateralAsset).aTokenAddress;
+    address debtATokenAddress = _getReserveData(repayParams.debtAsset).aTokenAddress;
+    if (repayParams.collateralAsset != repayParams.debtAsset) {
+      uint256 amounts0 = _getAmountsIn(
+        repayParams.useATokenAsFrom ? collateralATokenAddress : repayParams.collateralAsset,
+        repayParams.useATokenAsTo ? debtATokenAddress : repayParams.debtAsset,
+        repayParams.debtRepayAmount.add(premium),
+        repayParams.useEthPath
+      )[0];
+      require(amounts0 <= repayParams.collateralAmount, 'slippage too high');
+      uint256 feeAmount = amounts0.mul(FEE).div(FEE_DECIMALS);
+
+      if (repayParams.useATokenAsFrom) {
```

Comment:
Fee should always be transferred as AToken. Preferably straight from user to caller.